dialogue-core/src/main/java/com/palantir/dialogue/core/DialogueChannel.java

```diff
@@ -179,14 +188,27 @@ public DialogueChannel build() {
                 return new NeverThrowEndpointChannel(channel); // this must come last as a defensive backstop
             };
 
+            Supplier<Channel> stickyChannelSupplier = StickyEnpointChannels2.create(() -> {
+                LimitedChannel stickyLimitedChannel =
+                        StickyConcurrencyLimitedChannel.create(nodeSelectionChannel, cf.channelName());
+                Channel queueOverride = QueuedChannel.createForSticky(cf, stickyLimitedChannel);
```

Comment:
<just my own notes, not requesting any change>

This provides one queue per sticky session:

This means we get fairness between individual sticky sessions as well as requests which don't use sticky sessions, however in the worst case there may be many sessions in flight which makes it possible for one 'forced' request for each transaction to bypass the queue and limits. I think this is a good trade-off to avoid complexity.

The alternative is to associate queues with hosts so they're shared by all sticky sessions that have stuck to a given host. Each use is likely a bit different, so I like the idea of separate concurrency limits to punish 'bad' transactions without penalizing healthy transactions so much.

Reply:
> however in the worst case there may be many sessions in flight which makes it possible for one 'forced' request for each transaction to bypass the queue and limits.

So if you actually go back to the dominant resource fairness, and fairness in general, this is actually a perfectly valid strategy: by the way I implemented this, you only punch through the limits IFF you have zero other requests in flight. That means that in order to achieve min-max fairness you should absolutely let them through, because by definition they have the lowest utilisation of the available resources :)